### PR TITLE
fix(backend): prefer local copilot CLI over npx at launch

### PR DESCRIPTION
**Pull request by zeval**
Copilot launches paid an `npx -y @github/copilot` npm registry round-trip on every start — minutes behind a private registry — even when the `copilot` CLI was already installed; the lifecycle now runs the local binary when present.

## Important Changes

- New optional `NativeBinaryAgent` capability + `CommandOptions.PreferNativeBinary`; copilot's `BuildCommand` emits `copilot --acp` when preferred, else `npx`.
- `preferNativeBinary` reuses env-correct checks: standalone `exec.LookPath` (the reported local case), SSH preflight probe via metadata. Containerized runtimes (docker/sprites) keep `npx`.
- Re-probed every launch and not persisted, so an installed/removed binary is always reflected.

## Validation

- `go build ./...`
- `go test ./internal/agent/agents/` (144 pass) and `./internal/agent/runtime/lifecycle/` (668 pass), incl. new tests: copilot native-vs-npx command, `preferNativeBinary` across all runtimes with a real PATH lookup.
- `make fmt`, `golangci-lint run` on changed packages — clean.

## Possible Improvements

Low risk. Passthrough (`PassthroughCmd`) and one-shot inference (`InferenceConfig`) still use `npx`; copilot-only by design (claude-acp's launch package differs from its probed binary).

## Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have manually tested my changes and they work as expected.
- [ ] My changes have tests that cover the new functionality and edge cases.
- [ ] If my change touches UI files (`apps/web/`), I have added or updated Playwright e2e tests in `apps/web/e2e/` and verified them with `make test-e2e`.

<a href="https://cubic.dev/pr/kdlbs/kandev/pull/1384?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go [hunk 1]
```diff
@@ -600,6 +611,33 @@ func (r *SSHExecutor) preflightAgentBinary(ctx context.Context, client *ssh.Clie
...
+	if out == "" {
+		return false, nil
+	}
+	if req.Metadata == nil {
+		req.Metadata = make(map[string]interface{})
+	}
+	req.Metadata[MetadataKeyNativeBinary] = name
+	r.report(req.OnProgress, stepName, PrepareStepCompleted, out)
+	return true, nil
+}
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go]: Good catch — fixed in 8ea8101fd. The native-binary probe is now non-fatal: a transport error returns a miss (with a warn log) and falls through to the required npx command probe, which surfaces transport failures with proper context.

File: apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go [hunk 2]
```diff
@@ -600,6 +611,33 @@ func (r *SSHExecutor) preflightAgentBinary(ctx context.Context, client *ssh.Clie
...
+// declares one via NativeBinaryAgent). On a hit it records the binary name in
+// req.Metadata under MetadataKeyNativeBinary and returns found=true. A missing
+// binary returns found=false with no error so the caller falls back to the
+// default command. err is non-nil only on SSH transport failure.
+func (r *SSHExecutor) probeNativeBinary(ctx context.Context, client *ssh.Client, shell string, req *ExecutorCreateRequest, stepName string) (bool, error) {
+	nb, ok := req.AgentConfig.(agents.NativeBinaryAgent)
+	if !ok || nb.NativeBinaryName() == "" {
+		return false, nil
+	}
+	name := nb.NativeBinaryName()
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go]: Done in 8ea8101fd — split the type assertion from the name check so NativeBinaryName() is called once.

File: apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go [hunk 3]
```diff
@@ -600,6 +611,33 @@ func (r *SSHExecutor) preflightAgentBinary(ctx context.Context, client *ssh.Clie
...
+		return false, fmt.Errorf("ssh: probe %s on remote: %w", name, err)
+	}
+	if out == "" {
+		return false, nil
+	}
+	if req.Metadata == nil {
+		req.Metadata = make(map[string]interface{})
+	}
+	req.Metadata[MetadataKeyNativeBinary] = name
+	r.report(req.OnProgress, stepName, PrepareStepCompleted, out)
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/executor_ssh.go:648]: Good defensive catch — fixed in 6371e4675. Cleared MetadataKeyNativeBinary up front (before the probe) so a miss or transport error can't leave a stale value that flips selection to copilot --acp on an npx-only host.

File: apps/backend/internal/agent/runtime/lifecycle/manager_launch.go [hunk 4]
```diff
@@ -944,7 +946,9 @@ func (m *Manager) buildExecutionFromInstance(
 		execution.isResumedSession = true
 	}
 	execution.IsPassthrough = req.IsPassthrough
-	cmds := m.buildAgentCommand(req, profileInfo, agentConfig)
+	runtime := models.ExecutorType(req.ExecutorType).Runtime()
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/manager_launch.go]: Fixed in 9052f26cb — both call sites now pass execution.RuntimeName, so the runtime source is consistent.

File: apps/backend/internal/agent/runtime/lifecycle/manager_launch.go [hunk 5]
```diff
@@ -944,7 +946,9 @@ func (m *Manager) buildExecutionFromInstance(
...
+	preferNative := m.preferNativeBinary(agentConfig, runtime, execReq.Metadata)
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/runtime/lifecycle/manager_launch.go]: Fixed in 9052f26cb — buildExecutionFromInstance now uses the already-resolved execution.RuntimeName instead of re-deriving from req.ExecutorType, matching promoteWorkspaceExecution.

File: apps/backend/internal/agent/agents/copilot_acp.go [hunk 6]
```diff
@@ -79,7 +85,14 @@ func (a *CopilotACP) IsInstalled(ctx context.Context) (*DiscoveryResult, error)
 	return result, nil
 }
 
+// NativeBinaryName returns the standalone Copilot CLI name probed for in the
+// execution environment. See NativeBinaryAgent.
+func (a *CopilotACP) NativeBinaryName() string { return copilotNativeBinary }
+
 func (a *CopilotACP) BuildCommand(opts CommandOptions) Command {
+	if opts.PreferNativeBinary {
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/agents/copilot_acp.go:93]: Intentional. RuntimeConfig.Cmd has no production consumer (grep shows it's only referenced in tests); the launch path builds the agent command via BuildCommand. And by design containerized runtimes keep npx — PreferNativeBinary is only ever true for standalone/SSH — so Runtime().Cmd staying on npx is correct, not divergent. Leaving as-is.

File: apps/backend/internal/agent/settings/controller/agent_config.go [hunk 7]
```diff
@@ -214,6 +216,26 @@ func (c *Controller) PreviewAgentCommand(ctx context.Context, agentName string,
...
 }
 
+// previewPrefersNativeBinary reports whether the command preview should show a
+// NativeBinaryAgent's standalone CLI (e.g. "copilot --acp") instead of
+// `npx -y <pkg>`. The preview is not bound to an executor, so it assumes the
+// local (standalone) host — the default executor — and probes this machine's
+// PATH, mirroring the standalone branch of lifecycle.preferNativeBinary.
+// Containerized executors keep npx at launch regardless; the preview reflects
+// the common local case.
+func previewPrefersNativeBinary(agentConfig agents.Agent) bool {
```
[the remark this answers is not in the document]
Comment (zeval) [reply apps/backend/internal/agent/settings/controller/agent_config.go:226]: Intentional and documented in the function comment. The preview is not bound to an executor (profiles are executor-agnostic; the executor is chosen per run), so it can't know the launch runtime — and for SSH/remote the binary can't be probed at config time at all. It assumes the default standalone host and probes PATH, which is exactly the reported case (local copilot install showing npx). Containerized launches still use npx at runtime by design; the preview reflects the common local case. Making it fully executor-accurate would require plumbing a selected executor type into the preview endpoint — out of scope here. Resolving as intentional.
